### ISSUE TITLE
GLM4-base  bf16 int4量化微调之后，胡乱输出

### ISSUE THREAD

**Issue by pangpang-xuan** [THUDM/GLM-4#135]
### System Info / 系統信息

cuda 11.8 transformers=4.38.2 vllm=0.4.0 torch=2.1.0 python=3.10

### Information / 问题信息

- [ ] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

微调脚本如下：
torchrun --nnodes ${NODES} \
	--nproc_per_node ${NUM_GPUS} \
	--node_rank=${NODE_RANK} \
	--master_addr=${MASTER_ADDR} \
    --master_port=${MASTER_PORT} \
    src/train_bash.py \
    --deepspeed ${ds_config} \
    --stage sft \
    --do_train \
    --finetuning_type lora \
    --lora_target all \
    --lora_rank 64 \
    --model_name_or_path ${model_name_or_path} \
    --template ${template} \
    --dataset ${dataset} \
    --output_dir ${output_dir} \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 64 \
    --lr_scheduler_type cosine \
    --logging_steps 2 \
    --save_strategy epoch \
    --learning_rate 3e-4 \
    --num_train_epochs 5.0 \
    --warmup_ratio 0.1 \
    --plot_loss \
    --bf16 \
    --flash_attn \
    --seed 42 \
    --ddp_timeout 1800000 \
    --dataloader_num_workers 1 \
    --cutoff_len 2048 >> $OUTPUT_LOG 2>&1 \
    --quantization_bit 4
    
    zero2配置如下：
    {
  "train_batch_size": "auto",
  "train_micro_batch_size_per_gpu": "auto",
  "gradient_accumulation_steps": "auto",
  "gradient_clipping": "auto",
  "zero_allow_untested_optimizer": true,
  "fp16": {
    "enabled": "auto",
    "loss_scale": 0,
    "initial_scale_power": 16,
    "loss_scale_window": 1000,
    "hysteresis": 2,
    "min_loss_scale": 1
  },  
  "bf16": {
    "enabled": "auto"
  },
  "zero_optimization": {
    "stage": 2,
    "offload_optimizer": {
      "device": "cpu",
      "pin_memory": true
    },
    "offload_param": {
      "device": "cpu",
      "pin_memory": true
    },
    "allgather_partitions": true,
    "allgather_bucket_size": 5e8,
    "reduce_scatter": true,
    "reduce_bucket_size": 5e8,
    "overlap_comm": false,
    "contiguous_gradients": true
  }
}
    
    

### Expected behavior / 期待表现

期待输出：results = [\n\tSpatialEntity(\"LONDON\")\n]
实际输出：[gMASK] , LONDON\")\n] <|endoftext|>

**Comment by zRzRzRzRzRzRzR** [comment]
int4调不了的，只能BF16，而且，确保你的输出的时候解码成功skip special token
